CA-29 - Restrict authorized actions for EMR service role based on tags

hbarros-caylent opened the pull request:
Adds the map variable `abac_valid_tags`. It is a map(list(string)), the list of string contain valid values for each key.
The key and values of this map will dynamically create conditions to be used inside EMR Service Role Policies.
The examples were updated in order to reflect those changes.
A test case for the emr-iam module alone was added.


This PR links to [CA-31](https://github.com/Datatamer/terraform-aws-tamr-vm/pull/29) and [CA-33](https://github.com/Datatamer/terraform-aws-tamr-config/pull/15) and they all should be merged so that Epic CA-28 is done.

hbarros-caylent committed:
adds a simple test for IAM local module. links to CA-19

hbarros-caylent committed:
refactoring arn prefixes

hbarros-caylent committed:
Merge remote-tracking branch 'upstream/master' into refactor

hbarros-caylent committed:
first test with policy pasted from docs

hbarros-caylent committed:
Puts ResourceTags and RequestTags conditions w/ comments

hbarros-caylent committed:
removing unused resources and comments

hbarros-caylent committed:
adding variable abac_tags

hbarros-caylent committed:
terraform fmt

hbarros-caylent committed:
make lint terraform/docs

hbarros-caylent committed:
removing unused comments

hbarros-caylent committed:
reverting default value of core_instance_on_demand_count

hbarros-caylent committed:
refactors long arn strings

alexis-caylent commented on modules/aws-emr-iam/main.tf:
```diff
@@ -55,30 +57,57 @@ data "aws_iam_policy_document" "emr_service_policy_1" {
     resources = [
-      "*"
-    ]
+      "arn:aws:ec2:*:*:instance/*",
+      "arn:aws:ec2:*:*:volume/*",
+      "arn:aws:ec2:*:*:network-interface/*",
+      "arn:aws:ec2:*:*:launch-template/*"
```
Replace `"arn:aws:ec2:*:*:<RESOURCE_TYPE>/*"` with `"arn:${var.arn_partition}:ec2:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:<RESOURCE_TYPE>/*"`

hbarros-caylent replied:
Addressed with the refactoring

alexis-caylent commented on modules/aws-emr-iam/main.tf:
```diff
@@ -55,30 +57,57 @@ data "aws_iam_policy_document" "emr_service_policy_1" {
     ]
-    resources = ["*"]
+    resources = [
+      "arn:aws:ec2:*:*:instance/*",
+      "arn:aws:ec2:*:*:volume/*",
+      "arn:aws:ec2:*:*:network-interface/*",
+      "arn:aws:ec2:*:*:launch-template/*"
```
Replace `"arn:aws:ec2:*:*:<RESOURCE_TYPE>/*"` with `"arn:${var.arn_partition}:ec2:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:<RESOURCE_TYPE>/*"`

hbarros-caylent replied:
Addressed with the refactoring

alexis-caylent commented on variables.tf:
```diff
@@ -70,6 +70,14 @@ variable "tags" {
   default     = {}
 }
 
+variable "abac_tags" {
+  type        = map(string)
+  description = "Map of tags that will be used inside IAM Policies for restricting EMR Service Role access"
+  default = {
+    "tamr.com/role" = "emr"
+  }
+}
```
We need to decide whether we are going to set any tags by default. We want permissions to be as tight as possible out of the box, but we risk breaking things if the user forgets to tag the EMR subnet accordingly.

hbarros-caylent committed:
Better comments

hbarros-caylent committed:
replaces more string arns with the refactored strings

hbarros-caylent committed:
makes it optional for users to use abac_tags

hbarros-caylent committed:
fixes conditions for SpotRequests

hbarros-caylent committed:
Refactors abac_tags var from map(string) to map(list(string))

Before we tried using a map(string) that would indicate both the tags to use when tagging as the valid ones to insert into the IAM Polices. Which was not the right approach as the tags used for tagging is a map(string) and the condition tags is a map(list(string)), because we have a list of accepted values for each key.

hbarros-caylent committed:
Merges 6.0.0 into CA-29

hbarros-caylent committed:
more fixes from 3883482 and removes unused comments

hbarros-caylent committed:
adding abac tag condition for EMR managed security group

hbarros-caylent committed:
removing hardcoded vars in test_examples

hbarros-caylent committed:
terraform fmt

hbarros-caylent committed:
update version and changelog

hbarros-caylent committed:
make terraform/docs

souza-dan commented on test/create_module_emr-iam_test.go:
```diff
@@ -0,0 +1,33 @@
+package test
```
I'm assuming having the terratest test was helpful in your development, but including it here is slightly out of scope for this epic.

I'm fine with keeping it around though, this test is benign enough even if we haven't setup the integrations yet or refactor the format later it won't be a large change.

hbarros-caylent replied:
Yeah, I agree this would make more sense only if we had other tests merged already. I'm afraid though that the check from `continuous-integration/jenkins/pr-merge` won't merge because of *.go files check failing. Please look into the jenkins job and If that's the case let me know... so that we merge using the fastest route (maybe leave these out for now)

souza-dan commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
 
+## v6.1.0 - August 4th, 2021
+
+* Adds new variable `abac_valid_tags` to be used in IAM Policies conditions for creating EMR Resources using ABAC
```
Since the policy does change quite a bit, does upgrading from version v6.0.0 go fairly smoothly? Otherwise we may want to mark this as a major version.

hbarros-caylent replied:
They are safe since policy updates are done in-place. The new policies won't break existing clusters as all "old" statements were either kept or merged into the new ones. Also, in case the user doesn't pass `abac_valid_tags` conditions won't be created. 
The one problem I see happening here is if a user pass a map that is too long and that would exceed the limit of chars in a policy. 
I haven't tested the situation of breaking like this in an update but I think AWS API just wouldn't perform the update and keep the current policy working.
We can test those cases before proceeding just to be extra careful.

souza-dan reviewed (approved):
LGTM, updates to the policy looks solid!

souza-dan commented:
![image](https://user-images.githubusercontent.com/23456409/128527941-ff7e080c-f127-40f7-92a1-3cd921347166.png)

hbarros-caylent committed:
removes tests from this PR

hbarros-caylent committed:
Merge branch 'master' into CA-29

tamr-github-integrator commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on [GitHub release](https://github.com/Datatamer/terraform-aws-emr/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
